make MIR type checker handle a number of other cases

nikomatsakis opened the pull request:
The existing type checker was primarily used to verify types, but was skipping over a number of details. For example, it was not checking that the predicates on functions were satisfied and so forth. This meant that the NLL region checker was not getting a lot of the constraints it needed. This PR closes those gaps. It also includes a bit of refactoring for the way that we store region values, encapsulating the bit matrix over into its own module and improving the data structures in use.

This is mostly work by @spastorino being ported over from nll-master.

r? @arielb1 or @pnkfelix

bors commented:
:umbrella: The latest upstream changes (presumably #46640) made this pull request unmergeable. Please resolve the merge conflicts.

arielb1 commented on src/librustc_mir/transform/type_check.rs:
```diff
@@ -1110,21 +1151,123 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
+
+                    let ty_fn_ptr_from = tcx.mk_fn_ptr(fn_sig);
+
+                    if let Err(terr) = self.eq_types(ty_fn_ptr_from, ty, location.at_self()) {
```
I think typeck allows some sort of (e.g. higher-ranked) subtyping here, so you'll need to allow it too.

e.g.
```Rust
// this sig contains a binder
fn foo<'a>(_x: &'a ()) {}
fn bar<'b>() {
    foo as fn(&'b ()); // ... but this doesn't
}
```

nikomatsakis replied:
Hmm, that code type-checks. Perhaps because I haven't addressed previous comment yet.

arielb1 commented on src/test/compile-fail/mir_check_cast_reify.rs:
```diff
@@ -0,0 +1,52 @@
+    // During NLL region analysis, this will get renumbered to `typeof(foo::<'?0>)`
+    // where `'?0` is a new region variable.
+    //
+    // (Note that if `'a` on `foo` were early-bound, the type would be
```
... but 'a *is* early-bound. Did you mean late-bound?

nikomatsakis replied:
I did mean late-bound -- if it were late-bound, it wouldn't be part of the type of `foo`

nikomatsakis commented on src/librustc_mir/borrow_check/nll/universal_regions.rs:
```diff
@@ -493,7 +493,27 @@ impl<'cx, 'gcx, 'tcx> UniversalRegionsBuilder<'cx, 'gcx, 'tcx> {
+            // types.  What is the best way to handle this? Should we
+            // be checking something other than the type of the def-id
+            // to figure out what to do (e.g. the def-key?).
+            ty::TyUint(..) => {
```
It ICEs. I have later patches that do better.

nikomatsakis commented:
OK @arielb1 pushed some comments to address those.

nikomatsakis commented:
rebased, fixed a few minor things relating to changes in the underlying borrowck leading to slightly different errors, but I kept the separate commits in place

arielb1 commented:
Error message needs fixing:
```
[00:59:35] 
[00:59:35] ---- [compile-fail] compile-fail/nll/where_clauses_in_repeat_rvalue.rs stdout ----
[00:59:35] 	
[00:59:35] error: /checkout/src/test/compile-fail/nll/where_clauses_in_repeat_rvalue.rs:30: unexpected error: '30:17: 30:19: `x` does not live long enough [E0597]'
[00:59:35] 
[00:59:35] error: /checkout/src/test/compile-fail/nll/where_clauses_in_repeat_rvalue.rs:38: expected error not found: borrowed value does not live long enough [E0597]
[00:59:35] 
[00:59:35] error: 1 unexpected errors found, 1 expected errors not found
[00:59:35] status: exit code: 101
[00:59:35] command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/compile-fail/nll/where_clauses_in_repeat_rvalue.rs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/nll/where_clauses_in_repeat_rvalue.stage2-x86_64-unknown-linux-gnu" "-Crpath" "-O" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Z" "borrowck=mir" "-Z" "nll" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/nll/where_clauses_in_repeat_rvalue.stage2-x86_64-unknown-linux-gnu.aux" "-A" "unused"
[00:59:35] unexpected errors (from JSON output): [
[00:59:35]     Error {
[00:59:35]         line_num: 30,
[00:59:35]         kind: Some(
[00:59:35]             Error
[00:59:35]         ),
[00:59:35]         msg: "30:17: 30:19: `x` does not live long enough [E0597]"
[00:59:35]     }
[00:59:35] ]
[00:59:35] 
[00:59:35] not found errors (from test file): [
[00:59:35]     Error {
[00:59:35]         line_num: 38,
[00:59:35]         kind: Some(
[00:59:35]             Error
[00:59:35]         ),
[00:59:35]         msg: "borrowed value does not live long enough [E0597]"
[00:59:35]     }
[00:59:35] ]
[00:59:35] 
[00:59:35] thread '[compile-fail] compile-fail/nll/where_clauses_in_repeat_rvalue.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:1163:12
[00:59:35] 
[00:59:35] 
[00:59:35] failures:
[00:59:35]     [compile-fail] compile-fail/nll/where_clauses_in_repeat_rvalue.rs
[00:59:35] 
[00:59:35] test result: [31mFAILED(B[m. 2698 passed; 1 failed; 15 ignored; 0 measured; 0 filtered out
[00:59:35] 
```

nikomatsakis commented:
@arielb1 pushed an update to that error msg...let's see what travis thinks

arielb1 commented:
@bors r+

bors commented:
:pushpin: Commit 237dd41 has been approved by `arielb1`

<!-- @bors r=arielb1 237dd4121157943ed54a2e3f75b5849fadd8c7b4 -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 2, this pull request will be tested once the tree is reopened

bors commented:
:hourglass: Testing commit 237dd4121157943ed54a2e3f75b5849fadd8c7b4 with merge f03e06762ea6c4116c360185c0e9b6f209b76959...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5700), [status-travis](https://travis-ci.org/rust-lang/rust/builds/316243448?utm_source=github_status&utm_medium=notification)
Approved by: arielb1
Pushing f03e06762ea6c4116c360185c0e9b6f209b76959 to master...
